src/js/wrap.cpp

```diff
@@ -81,28 +82,73 @@ namespace ccf::js
     }
   }
 
+  static int js_custom_interrupt_handler(JSRuntime* rt, void* opaque)
+  {
+    UntrustedHostTime* time = reinterpret_cast<UntrustedHostTime*>(opaque);
+    auto now = ccf::get_enclave_time();
+    auto elapsed_time = now - time->start_time;
+    auto elapsed_ms =
+      std::chrono::duration_cast<std::chrono::milliseconds>(elapsed_time);
+    if (elapsed_ms.count() >= time->max_execution_time.count())
+    {
+      LOG_INFO_FMT("JS execution has timed out");
+      time->request_timed_out = true;
+      return 1;
+    }
+    else
+    {
+      return 0;
+    }
+  }
+
+  JSWrappedValue Context::call(
+    const JSWrappedValue& f, const std::vector<js::JSWrappedValue>& argv) const
+  {
+    std::vector<JSValue> argvn;
+    argvn.reserve(argv.size());
+    for (auto& a : argv)
+    {
+      argvn.push_back(a.val);
+    }
+
+    const auto curr_time = ccf::get_enclave_time();
+    host_time->start_time = curr_time;
+    host_time->max_execution_time = max_execution_time;
+    JS_SetInterruptHandler(js_run_time, js_custom_interrupt_handler, host_time);
+
+    return W(JS_Call(ctx, f, JS_UNDEFINED, argv.size(), argvn.data()));
+  }
+
   Runtime::Runtime(kv::Tx* tx)
   {
     rt = JS_NewRuntime();
     if (rt == nullptr)
     {
       throw std::runtime_error("Failed to initialise QuickJS runtime");
     }
-    size_t stack_size = 1024 * 1024;
-    size_t heap_size = 100 * 1024 * 1024;
+    size_t stack_size = default_stack_size;
+    size_t heap_size = default_heap_size;
+
     const auto jsengine = tx->ro<ccf::JSEngine>(ccf::Tables::JSENGINE);
     const std::optional<JSRuntimeOptions> js_runtime_options = jsengine->get();
 
     if (js_runtime_options.has_value())
     {
       heap_size = js_runtime_options.value().max_heap_bytes;
       stack_size = js_runtime_options.value().max_stack_bytes;
+      max_execution_time = std::chrono::milliseconds{
+        js_runtime_options.value().max_execution_time_ms};
     }
 
     JS_SetMaxStackSize(rt, stack_size);
     JS_SetMemoryLimit(rt, heap_size);
   }
 
+  Runtime::~Runtime()
+  {
+    JS_FreeRuntime(rt);
```

Comment:
Did we not free the runtime before? It's very surprising this wasn't caught by the performance tests, considering we create a new runtime every request.

Reply:
I see, this was moved.